### PR TITLE
Fix the rule of port allocate when executing -p multiple times for one command

### PR DESCRIPTION
**Pull request by tces1**
#2256 
Fix public ports always assign to the same port numbe issue

### CODE REVIEW

File: pkg/portutil/port_allocate_linux.go [hunk 1]
```diff
@@ -25,10 +25,13 @@ import (
 
 const (
 	// This port range is compatible with Docker, FYI https://github.com/moby/moby/blob/eb9e42a09ee123af1d95bf7d46dd738258fa2109/libnetwork/portallocator/portallocator_unix.go#L7-L12
-	allocateStart = 49153
 	allocateEnd   = 60999
 )
 
+var (
+	allocateStart uint64 = 49153
+)
```
Comment (AkihiroSuda) [inline pkg/portutil/port_allocate_linux.go:33]: Still conflicts if you run `nerdctl run` commands twice?

Comment (tces1) [reply pkg/portutil/port_allocate_linux.go:33]: I checked the previous commit and someone someone solved the problem of repeated executions of nerdctl run always allocating 49153 by reading the iptables nat table to determine which are the ports that have already been allocated. So using it this way here does not reintroduce the problem of always allocating the same port for multiple executions, what is solved here is the rule of port allocate when executing -p multiple times for one command

File: pkg/portutil/port_allocate_linux.go [hunk 2]
```diff
@@ -96,6 +99,7 @@ func portAllocate(protocol string, ip string, count uint64) (uint64, uint64, err
 			}
 		}
 		if needReturn {
+			allocateStart = start + count
 			return start, start + count - 1, nil
```
[the remark this answers is not in the document]
Comment (tces1) [reply pkg/portutil/port_allocate_linux.go:103]: I think this rule just follow the docker client

File: pkg/portutil/port_allocate_linux.go (hunk 1 above)
Comment (AkihiroSuda) [reply pkg/portutil/port_allocate_linux.go:33]: - https://github.com/containerd/nerdctl/pull/2169

@vsiravar Could you take a llook?

Comment (vsiravar) [reply pkg/portutil/port_allocate_linux.go:33]: >  what is solved here is the rule of port allocate when executing -p multiple times for one command

This doesn't seem to be the case from my local test from this PR. 
```
$ sudo nerdctl run -d -p 80 -p 90 nginx
$ sudo nerdctl ps -a
CONTAINER ID    IMAGE                                       COMMAND                   CREATED               STATUS                     PORTS                                           NAMES
8d80a3e262d2    docker.io/library/nginx:latest              "/docker-entrypoint.…"    About a minute ago    Up                         0.0.0.0:49153->80/tcp, 0.0.0.0:49153->90/tcp    nginx-8d80a

Comment (fahedouch) [reply pkg/portutil/port_allocate_linux.go:33]: > what is solved here is the rule of port allocate when executing -p multiple times for one command

you have to rename your PR to reflect this in the title plz

Comment (tces1) [reply pkg/portutil/port_allocate_linux.go:33]: > > what is solved here is the rule of port allocate when executing -p multiple times for one command
> 
> you have to rename your PR title to reflect the above in title

Ok, I have renamed it

Comment (tces1) [reply pkg/portutil/port_allocate_linux.go:33]: > > what is solved here is the rule of port allocate when executing -p multiple times for one command
> 
> This doesn't seem to be the case from my local test from this PR.
> 
> ```
> $ sudo nerdctl run -d -p 80 -p 90 nginx
> $ sudo nerdctl ps -a
> CONTAINER ID    IMAGE                                       COMMAND                   CREATED               STATUS                     PORTS                                           NAMES
> 8d80a3e262d2    docker.io/library/nginx:latest              "/docker-entrypoint.…"    About a minute ago    Up                         0.0.0.0:49153->80/tcp, 0.0.0.0:49153->90/tcp    nginx-8d80a
> ```

Did you pull my changes and compile them, I looked and I didn't miss any changes,
I have tested locally with multiple ports and it seems to be working fine and is currently in use at work.

bfd5c6f92869   docker:dind    "/sbin/tini -- docke…"    4 seconds ago    Up        0.0.0.0:49153->7222/tcp, 0.0.0.0:49154->13022/tcp, 0.0.0.0:49155->1847/tcp, 0.0.0.0:49156->7024/tcp, 0.0.0.0:49157->8222/tcp, 0.0.0.0:49158->14022/tcp, 0.0.0.0:49159->3830/tcp, 0.0.0.0:49160->8024/tcp, 0.0.0.0:49161->9222/tcp, 0.0.0.0:49162->15022/tcp, 0.0.0.0:49163->1846/tcp, 0.0.0.0:49164->9024/tcp, 0.0.0.0:49165->10222/tcp, 0.0.0.0:49166->16022/tcp, 0.0.0.0:49167->1845/tcp, 0.0.0.0:49168->10024/tcp, 0.0.0.0:49169->3022/tcp, 0.0.0.0:49170->1848/tcp, 0.0.0.0:49171->4022/tcp, 0.0.0.0:49172->4830/tcp, 0.0.0.0:49173->61000/tcp, 0.0.0.0:49174->62000/tcp, 0.0.0.0:49175->62001/tcp, 0.0.0.0:49176->62002/tcp, 0.0.0.0:49177->62003/tcp, 0.0.0.0:49178->3255/tcp   boards_1_mh2roh_root

Comment (tces1) [reply pkg/portutil/port_allocate_linux.go:33]: image

**Comment by tces1** [comment]
HI, is any comments about this pr?

File: pkg/portutil/port_allocate_linux.go [hunk 3]
```diff
@@ -96,6 +99,7 @@ func portAllocate(protocol string, ip string, count uint64) (uint64, uint64, err
 			}
 		}
 		if needReturn {
+			allocateStart = start + count
```
Comment (AkihiroSuda) [inline pkg/portutil/port_allocate_linux.go:102]: Can we have an integration test?

**Comment by AkihiroSuda** [comment]
lint is failing

**Comment by AkihiroSuda** [comment]
Please squash commits

**Comment by tces1** [comment]
> Please squash commits

Thanks for the heads up, this is my first pr to a formal project and some of the features are not very familiar, did I succeed with the squash operation please?

**Comment by AkihiroSuda** [comment]
> > Please squash commits
> 
> Thanks for the heads up, this is my first pr to a formal project and some of the features are not very familiar, did I succeed with the squash operation please?

You can remove "Fix golint issue" from the commit message.
The "Signed-off-by" line should be in the last.

**Comment by tces1** [comment]
> > > Please squash commits
> > 
> > 
> > Thanks for the heads up, this is my first pr to a formal project and some of the features are not very familiar, did I succeed with the squash operation please?
> 
> You can remove "Fix golint issue" from the commit message. The "Signed-off-by" line should be in the last.

Thanks, I have modified by your tips.

File: pkg/portutil/port_allocate_linux.go (hunk 3 above)
Comment (tces1) [reply pkg/portutil/port_allocate_linux.go]: Yes, we can

**Review by AkihiroSuda — APPROVED** [review]
Thanks, I'm merging this, but please consider adding an integration test

**Pull request merged**